src/engine/idTable/CompressedExternalIdTable.h

```diff
@@ -805,6 +829,84 @@ class CompressedExternalIdTableSorter
       return blockSizeForOutput;
     }
   };
+
+  template <size_t N, typename RowGenVectorType>
+  std::optional<IdTableStatic<N>> generateBlocks(
+      std::shared_ptr<SortState> state, RowGenVectorType& rowGenerators,
+      size_t blockSizeOutput, const auto& comp) {
+    if (!state->initialized_) {
+      state->initialized_ = true;
+      for (auto& gen : rowGenerators) {
+        state->priorityQueue_.emplace_back(gen.begin(), gen.end());
+      }
+      ql::ranges::make_heap(state->priorityQueue_, comp);
+      state->result_ = IdTableStatic<NumStaticCols>(this->writer_.numColumns(),
+                                                    this->writer_.allocator());
+      state->result_.reserve(blockSizeOutput);
+    }
+    while (!state->priorityQueue_.empty()) {
+      ql::ranges::pop_heap(state->priorityQueue_, comp);
+      auto& min = state->priorityQueue_.back();
+      state->result_.push_back(*min.first);
+      ++(min.first);
+      if (min.first == min.second) {
+        state->priorityQueue_.pop_back();
+      } else {
+        ql::ranges::push_heap(state->priorityQueue_, comp);
+      }
+      if (state->result_.size() >= blockSizeOutput) {
+        state->numPopped_ += state->result_.numRows();
+        auto resultBlock = std::move(state->result_).template toStatic<N>();
+        state->result_ = IdTableStatic<NumStaticCols>(
+            this->writer_.numColumns(), this->writer_.allocator());
+        state->result_.reserve(blockSizeOutput);
+        return resultBlock;
+      }
+    }
+    if (!state->result_.empty()) {
+      state->numPopped_ += state->result_.numRows();
+      auto resultBlock = std::move(state->result_).template toStatic<N>();
+      state->result_ = IdTableStatic<NumStaticCols>(this->writer_.numColumns(),
+                                                    this->writer_.allocator());
+      state->result_.clear();
+      AD_CORRECTNESS_CHECK(state->numPopped_ == this->numElementsPushed_);
+      return resultBlock;
+    }
+    return std::nullopt;
+  }
+
+  template <size_t N>
+  std::optional<IdTableStatic<N>> generateBlocksFromMemory(
+      std::optional<size_t> blocksize, bool& emptyYielded, size_t& currentPos) {
+    auto& block = this->currentBlock_;
+    if (block.empty()) {
+      return std::nullopt;
+    }
+    const auto blocksizeOutput = blocksize.value_or(block.numRows());
+    if (block.numRows() <= blocksizeOutput) {
+      if (emptyYielded) {
+        return std::nullopt;
+      }
+      emptyYielded = true;
```

Comment:
We have this pattern many times for initialization flags. It can be
`if (!std::exchange(emptyYielded, true) { return nullopt;}`
(exchange returns the old value).